### PR TITLE
feat: Multi chart

### PR DESCRIPTION
**Pull request by pghildiyal**
# Description

Devtron should be able to provide multiple chart types for applications. Currently only Rollout type charts are supported by at times users need other types for eg Job type or knative type. This release will allow different chart  types to be added to the repository of Devtron

Fixes # devtron-labs/devtron#949

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Tested by creating new applications and adding environment overrides.

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/deploymentConfig/DeploymentConfig.tsx [hunk 1]
```diff
@@ -31,19 +32,20 @@ export function OptApplicationMetrics({ currentVersion, onChange, opted, focus =
...
-        <DeploymentConfigForm respondOnSuccess={respondOnSuccess} />
+        <DeploymentConfigForm respondOnSuccess={respondOnSuccess} isUnSet={isUnSet} />
     </div>
 }
 
-function DeploymentConfigForm({ respondOnSuccess }) {
-    const [chartVersions, setChartVersions] = useState<{ id: number, version: string; }[]>(null)
-    const [selectedChart, selectChart] = useState<{ id: number, version: string; }>(null)
+function DeploymentConfigForm({ respondOnSuccess, isUnSet }) {
+    const [chartVersions, setChartVersions] = useState<{ id: number, version: string, name: string; }[]>([])
```
Comment (manish-agrawal-ai) [inline src/components/deploymentConfig/DeploymentConfig.tsx:46]: chartVersions variable name is confusing here. (It should be charts)

File: src/components/deploymentConfig/DeploymentConfig.tsx [hunk 2]
```diff
@@ -167,36 +169,99 @@ function DeploymentConfigForm({ respondOnSuccess }) {
         }
     }
     const appMetricsEnvironmentVariableEnabled = window._env_ && window._env_.APPLICATION_METRICS_ENABLED;
+    let uniqueCharts = new Map<string, boolean>();
+    let chartNames = chartVersions.filter(cv => {if (uniqueCharts.get(cv.name)) { return false } else { uniqueCharts.set(cv.name, true);  return true;}});
+    let filteredChartVersions = selectedChart? chartVersions.filter(cv => cv.name == selectedChart.name) : [];
```
Comment (manish-agrawal-ai) [inline src/components/deploymentConfig/DeploymentConfig.tsx:174]: variable name should be filteredCharts

File: src/components/deploymentConfig/DeploymentConfig.tsx [hunk 3]
```diff
@@ -167,36 +169,99 @@ function DeploymentConfigForm({ respondOnSuccess }) {
         }
     }
     const appMetricsEnvironmentVariableEnabled = window._env_ && window._env_.APPLICATION_METRICS_ENABLED;
+    let uniqueCharts = new Map<string, boolean>();
+    let chartNames = chartVersions.filter(cv => {if (uniqueCharts.get(cv.name)) { return false } else { uniqueCharts.set(cv.name, true);  return true;}});
```
Comment (manish-agrawal-ai) [inline src/components/deploymentConfig/DeploymentConfig.tsx:173]: variable name assignment is irrelevant here.

File: src/components/deploymentConfig/DeploymentConfig.tsx [hunk 4]
```diff
@@ -167,36 +169,99 @@ function DeploymentConfigForm({ respondOnSuccess }) {
...
+                                        width: '100%'
+                                    })
+                                },
+                            }}
+                            onChange={(selected) => selectChart(selected as { id: number, version: string, name: string })}
+                        />
+                    </div>
+                </div>
+                <div className="deploymentConfig__warning flex fs-12 left pt-4 m-b-16">
+                    { isUnSet ? (
```
Comment (manish-agrawal-ai) [inline src/components/deploymentConfig/DeploymentConfig.tsx:254]: Can be simplified to :

              { isUnSet && 
                        <>
                        <Warn className="icon-dim-16 mr-4 " />
                        Chart type cannot be changed once saved.
                        </>
                    }

No ternary operation is needed

File: src/components/deploymentConfig/DeploymentConfig.tsx [hunk 5]
```diff
@@ -167,36 +169,99 @@ function DeploymentConfigForm({ respondOnSuccess }) {
...
+                                    width: '100%'
+                                }),
+                                option: (base, state) => {
+                                    return ({
+                                        ...base,
+                                        color: 'var(--N900)',
+                                        backgroundColor: state.isFocused ? 'var(--N100)' : 'white',
+                                    })
+                                },
+                                container: (base, state) => {
```
Comment (manish-agrawal-ai) [inline src/components/deploymentConfig/DeploymentConfig.tsx:242]: This is new (container was not there before. I assume this has no impact)

File: src/components/deploymentConfig/DeploymentConfig.tsx [hunk 6]
```diff
@@ -167,36 +169,99 @@ function DeploymentConfigForm({ respondOnSuccess }) {
...
+                            value={selectedChart}
+                            components={{
+                                IndicatorSeparator: null
+                            }}
+                            styles={{
+                                control: (base, state) => ({
+                                    ...base,
+                                    boxShadow: 'none',
+                                    border: `solid 1px var(--B500)`,
+                                    width: '100%'
```
Comment (manish-agrawal-ai) [inline src/components/deploymentConfig/DeploymentConfig.tsx:233]: I assume width: '100%' has no impact. as this was not there before

File: src/components/deploymentConfig/DeploymentConfig.tsx [hunk 7]
```diff
@@ -31,19 +32,20 @@ export function OptApplicationMetrics({ currentVersion, onChange, opted, focus =
...
     </div>
 }
 
-function DeploymentConfigForm({ respondOnSuccess }) {
-    const [chartVersions, setChartVersions] = useState<{ id: number, version: string; }[]>(null)
-    const [selectedChart, selectChart] = useState<{ id: number, version: string; }>(null)
+function DeploymentConfigForm({ respondOnSuccess, isUnSet }) {
+    const [chartVersions, setChartVersions] = useState<{ id: number, version: string, name: string; }[]>([])
+    const [selectedChart, selectChart] = useState<{ id: number, version: string, name: string; }>(null)
+    const [selectedChartRefId, setSelectedChartRefId] = useState(0)
```
Comment (manish-agrawal-ai) [inline src/components/deploymentConfig/DeploymentConfig.tsx:48]: setSelectedChartRefId - this usestate is unused

File: src/components/deploymentConfig/DeploymentConfig.tsx [hunk 8]
```diff
@@ -167,36 +169,99 @@ function DeploymentConfigForm({ respondOnSuccess }) {
...
+                                    })
+                                },
+                                container: (base, state) => {
+                                    return ({
+                                        ...base,
+                                        width: '100%'
+                                    })
+                                },
+                            }}
+                            onChange={(selected) => selectChart(selected as { id: number, version: string, name: string })}
```
Comment (manish-agrawal-ai) [inline src/components/deploymentConfig/DeploymentConfig.tsx:211]: On changing, we are selecting the chart. which is wrong IMO.

**Review by manish-agrawal-ai — COMMENTED** [review]
check comments

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 1 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:46]: done

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 7 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:48]: unused deleted

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 2 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:174]: done

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 4 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:254]: done

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 6 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:233]: not needed, removed

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 5 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:242]: this is correct

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 3 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:173]: for now lets keep it

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 7 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:48]: fixed, using it now

File: src/components/deploymentConfig/DeploymentConfig.tsx (hunk 8 above)
Comment (pghildiyal) [reply src/components/deploymentConfig/DeploymentConfig.tsx:211]: fixed the logic

File: src/components/EnvironmentOverride/DeploymentTemplateOverride.tsx [hunk 9]
```diff
@@ -215,6 +215,11 @@ function DeploymentTemplateOverrideForm({ state, handleOverride, dispatch, initi
                     onClick={handleOverride}
                     type="deployment template"
                 />
+                <div className="form__row">
```
Comment (manish-agrawal-ai) [inline src/components/EnvironmentOverride/DeploymentTemplateOverride.tsx:218]: Lets say I have selected Chart type as "Rollout" while setting global deployment template. 
Now Can I select "Cron Job" chart ref Id while overriding on environment? 
If Not, I can not see any handling for that.

**Review by manish-agrawal-ai — COMMENTED** [review]
check comment

File: src/components/EnvironmentOverride/DeploymentTemplateOverride.tsx (hunk 9 above)
Comment (pghildiyal) [reply src/components/EnvironmentOverride/DeploymentTemplateOverride.tsx:219]: fixed

File: src/components/EnvironmentOverride/DeploymentTemplateOverride.tsx [hunk 10]
```diff
@@ -206,6 +206,7 @@ function DeploymentTemplateOverrideForm({ state, handleOverride, dispatch, initi
         }
     }
     const appMetricsEnvironmentVariableEnabled = window._env_ && window._env_.APPLICATION_METRICS_ENABLED;
+    const name = state.charts.get(state.data.globalChartRefId)?.name;
```
Comment (manish-agrawal-ai) [inline src/components/EnvironmentOverride/DeploymentTemplateOverride.tsx:209]: rename this variable to chartName :P

**Review by manish-agrawal-ai — COMMENTED** [review]
please check comment

**Review by manish-agrawal-ai — APPROVED** [review]
approved

**Pull request merged**